### PR TITLE
Add fromHexString, fromHexStringAsRange and isHexString

### PR DESCRIPTION
**Pull request by 0xEAB**
Rebased & squashed #8523

### CODE REVIEW

**Comment by 0xEAB** [comment]
Okay, so I’ve added a few code-style refinements.

File: std/digest/package.d [hunk 1]
```diff
@@ -1212,3 +1212,137 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
...
+{
+    import std.ascii : isDigit, isAlpha, toLower;
+
+    if (hexDigit.isDigit)
+    {
+        return hexDigit - '0';
+    }
+    else if (hexDigit.isAlpha)
+    {
+        return hexDigit.toLower - 'a' + 10;
```
Comment (thewilsonator) [inline std/digest/package.d:1293]: what about upper case hex? what about non-hex alpha characters?

**Review by thewilsonator — COMMENTED** [review]
This doesn't appear to do any input validation. 

Does this fix an issue? If so please retitle the committee message appropriately, if not please add a changelog entry.

File: std/digest/package.d [hunk 2]
```diff
@@ -1212,3 +1212,137 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
...
+{
+    T hex;
+
+    size_t length;
+    ubyte front;
+    bool empty;
+
+    this(T hex)
+    {
+        if (hex.length > 0)
```
Comment (thewilsonator) [inline std/digest/package.d:1308]:
```
if (hex.length == 0)
{
    empty = true;
    length = 0;
    return;
}
```

**Comment by 0xEAB** [comment]
> Does this fix an issue?

I don’t think so. The original PR doesn’t mention anything in that regard. I’ll add a changelog entry.

File: std/digest/package.d (hunk 1 above)
Comment (0xEAB) [reply std/digest/package.d:1293]: > what about upper case hex?

Isn’t this covered by the `.toLower` call?

Comment (0xEAB) [reply std/digest/package.d:1293]: > what about non-hex alpha characters?

Yeah, the condition should prolly check for `isHexDigit`.

Comment (thewilsonator) [reply std/digest/package.d:1293]: > Isn’t this covered by the .toLower call?

I missed that.

**Comment by 0xEAB** [comment]
> @thewilsonator  
> This doesn't appear to do any input validation.

- I’ve added a validation function:
    <https://github.com/dlang/phobos/blob/9ff3ac3e43346e266e4fef51bd9d8e86f80dd48d/std/digest/package.d#L1227>
- The higher level function calls it and throws in case of failure:
    <https://github.com/dlang/phobos/blob/9ff3ac3e43346e266e4fef51bd9d8e86f80dd48d/std/digest/package.d#L1303-L1310>
- The lower level function is documented to require pre-validated input:
    <https://github.com/dlang/phobos/blob/9ff3ac3e43346e266e4fef51bd9d8e86f80dd48d/std/digest/package.d#L1260-L1261>
    - In case of invalid input, it crashes the program:
        <https://github.com/dlang/phobos/blob/9ff3ac3e43346e266e4fef51bd9d8e86f80dd48d/std/digest/package.d#L1382>

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**